File: frontend/src/types/role.ts
export interface Permission {
  id: string;
  name: string;
  resource: string;
  action: string;
  description?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface Role {
  id: string;
  name: string;
  description?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  permissions: Permission[];
}

export interface CreateRoleRequest {
  name: string;
  description?: string;
  permission_ids?: string[];
}

export interface UpdateRoleRequest {
  name?: string;
  description?: string;
  is_active?: boolean;
  permission_ids?: string[];
}

export interface CreatePermissionRequest {
  name: string;
  resource: string;
  action: string;
  description?: string;
}

export interface UpdatePermissionRequest {
  name?: string;
  resource?: string;
  action?: string;
  description?: string;
  is_active?: boolean;
}

export interface UserRoleInfo {
  role_id?: string;
  role_name?: string;
  permissions: string[];
}

export interface RoleFilters {
  search: string;
  status: 'active' | 'inactive' | 'all';
}

export interface PermissionFilters {
  search: string;
  resource?: string;
  action?: string;
  status: 'active' | 'inactive' | 'all';
}

File: backend/app/schemas/role.py
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field
from uuid import UUID


class PermissionBase(BaseModel):
    name: str = Field(..., description="Nome único da permissão")
    resource: str = Field(..., description="Recurso (ex: products, clients)")
    action: str = Field(..., description="Ação (ex: view, create, edit, delete)")
    description: Optional[str] = Field(None, description="Descrição da permissão")


class Permission(PermissionBase):
    id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class PermissionCreate(PermissionBase):
    pass


class PermissionUpdate(BaseModel):
    name: Optional[str] = None
    resource: Optional[str] = None
    action: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None


class RolePermissionBase(BaseModel):
    permission_id: UUID
    granted: bool = True


class RolePermission(RolePermissionBase):
    id: UUID
    role_id: UUID
    permission: Permission
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class RoleBase(BaseModel):
    name: str = Field(..., description="Nome único do papel")
    description: Optional[str] = Field(None, description="Descrição do papel")


class Role(RoleBase):
    id: UUID
    is_active: bool
    created_at: datetime
    updated_at: datetime
    permissions: List[Permission] = []
    
    class Config:
        from_attributes = True


class RoleCreate(RoleBase):
    permission_ids: Optional[List[UUID]] = []


class RoleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None
    permission_ids: Optional[List[UUID]] = None


class UserRoleInfo(BaseModel):
    role_id: Optional[UUID] = None
    role_name: Optional[str] = None
    permissions: List[str] = []  # Lista de permissões no formato "resource:action"